### PR TITLE
Add ammonia production

### PR DESCRIPTION
**Pull request by AtomicFox556**
#### Summary
Content "Add ammonia production"

#### Purpose of change
Add a way to produce ammonia.

#### Describe the solution
Make use of Haber-Borsch process for producing ammonia, adding the necessary items to do it. They can be used as standalone items, as appliances, or mounted on a vehicle to draw power from it.
Nitrogen membrane filter was added as an uncraftable key part to make it possible to capture nitrogen from air, based on this: https://www.generon.com/product/nitrogen-membrane-separators-nitrogen-separation-membranes/
Compressor item was also added, as there doesn't appear to be an easy way to make one from scratch (they appear to be somewhat similar to internal combustion engines in their construction). It's also now used for making condensor coil which would in turn be found in fridges.

#### Describe alternatives you've considered
Just requiring a pressure cooker for this pressure-sensitive reaction, but apparently the pressure required is well above maximum pressure that a pressure cooker can contain.

#### Testing
Launched the game with changes, made sure that no errors happen and that newly added vehicle-mounted items are available as tools for producing ammonia.

#### Additional context
Suggestions are welcome regarding where nitrogen membrane filters could be found commonly; I couldn't think of anything.

### CODE REVIEW

File: data/json/vehicleparts/utilities.json [hunk 1]
```diff
@@ -263,6 +263,76 @@
...
+      { "item": "pipe", "count": [ 1, 2 ] },
+      { "item": "compressor", "prob": 25 }
+    ]
+  },
+  {
+    "id": "veh_ammonia_machine",
+    "type": "vehicle_part",
+    "name": { "str": "mounted ammonia machine" },
+    "item": "ammonia_machine",
+    "description": "A huge makeshift appliance which produces ammonia via Haber–Bosch process from hydrogen and nitrogen, powered by the vehicle's electrical system.",
```
Comment (NetSysFire) [inline data/json/vehicleparts/utilities.json:306]:
```suggestion
    "description": "A huge makeshift appliance which produces ammonia via the Haber–Bosch process from hydrogen and nitrogen, powered by the vehicle's electrical system.",
```

File: data/json/furniture_and_terrain/appliances.json [hunk 2]
```diff
@@ -148,6 +148,26 @@
...
+    "item": "ammonia_machine",
+    "flags": [ "OBSTACLE", "APPLIANCE" ],
+    "copy-from": "veh_ammonia_machine"
+  },
+  {
+    "type": "vehicle_part",
+    "id": "ap_nitrogen_generator",
+    "name": { "str": "nitrogen generator" },
+    "location": "structure",
+    "description": "A large makeshift appliance which captures nitrogen from air for use in chemistry projects, with a power cord to connect it to an electric grid.",
```
Comment (NetSysFire) [inline data/json/furniture_and_terrain/appliances.json:166]:
```suggestion
    "description": "A large makeshift appliance which captures nitrogen from the air for use in chemistry projects, with a power cord to connect it to an electric grid.",
```

**Comment by Kamejeir** [comment]
This nitrogen filter... I visited the page and read "...they will generate a nitrogen stream that is suitable for a wide range of industries including beverage, laboratory, food, controlled atmosphere, pharmaceutical, chemical, textile, heat treatment and electronics."

I notice that it says "nitrogen stream", implying it isn't a batch process and therefore doesn't *have* to be part of a bottling facility.
Therefore I find it likely to be attached to a machine in the facilities harboring the processes mentioned. Maybe even an existing Haber-Bosch machine (which sadly isn't likely going to be found at farms).

As a result, the problem now lies in finding and dismantling a relevant industrial machine.

**Comment by AtomicFox556** [comment]
Further improved the process by replacing brine with sodium hydroxide solution for electrolysis, as it doesn't get used up during electrolysis and doesn't produce chlorine that is toxic.

So now you just need some salt (to replace salt that is used up for air drying) and water to make use of this process, and salt isn't used up as you can recover it by boiling off the salt water byproduct.

**Comment by PatrikLundell** [comment]
Based on the description (testing part), this thing seems to be vehicle mounted and it's also described as huge. Wouldn't this be a candidate for something that's available as a fixed appliance only (i.e. to appliance grids, but not in vehicles)?

**Comment by AtomicFox556** [comment]
> Based on the description (testing part), this thing seems to be vehicle mounted and it's also described as huge. Wouldn't this be a candidate for something that's available as a fixed appliance only (i.e. to appliance grids, but not in vehicles)?

Right now it's available as both; besides free space limitations I can't really think of anything that could prevent this thing from taking power from a vehicle. It isn't *that* huge either; its volume (90 L) is based on components used to make it, so even if it wasn't possible to mount it to a vehicle or even use it as a standalone item, could just place it into a vehicular cargo space and place it on the ground any time you'd need to make some ammonia; mounting it to a vehicle right away pretty much just saves time and effort.

**Comment by Fris0uman** [comment]
Vehicle part don't just take power from a vehicle they are structurally part of the vehicle and can run while the vehicle is moving. Would it make sense to have this thing function on a moving vehicle? Now that we have static appliances we need to stop sticking onto vehicle things that have no reason to go there.

**Comment by AtomicFox556** [comment]
> Vehicle part don't just take power from a vehicle they are structurally part of the vehicle and can run while the vehicle is moving. Would it make sense to have this thing function on a moving vehicle? Now that we have static appliances we need to stop sticking onto vehicle things that have no reason to go there.

It should function fine as well, but it'd be very difficult to use while the vehicle is moving (in the sense that anything you try to manipulate will get thrown around if the vehicle is moving or turning really fast).

I found no info regarding how resistant the electrolysis unit would be to being moved; my understanding is that particularly violent movement would at most cause temporary disruption until the solution settles, as long as the container is enclosed to prevent spilling and is not turned sideways or upside down.

**Review by kevingranade — CHANGES_REQUESTED** [review]
This isn't remotely representative of the Haber-Bosch process.

First off, a (very generous) power consumption rate is 100 GJ tNH₃−1 which is cited as the efficiency level of a large scale Haber-Bosch process in 1930. https://www.sciencedirect.com/topics/engineering/haber-bosch-process

If my math is correct, that means your 8-charge (992g) ammonia recipe should cost about 99,200,000 J of energy (that's actually heat energy, meaning the electrical cost would be even higher), not 500.

This is ignoring the problems of scaling (Haber-Bosch notoriously requires large reactors, high flow rates, high pressures, high temperatures, etc) that mean it's not at all practical to even drive this reaction electrically with the catalysts and gear you have outlined here and the fact that hydrogen electrolysis is also immensely inefficient without exotic catalysts.

At the end of the day, I'm expecting an overall conversion rate worse than 200 GJ / tNH3, meaning you're consuming well over 200 million battery charges per batch, as well as a much more robust and difficult crafting recipe for the machinery as well.

**Comment by AtomicFox556** [comment]
> This isn't remotely representative of the Haber-Bosch process.
> 
> First off, a (very generous) power consumption rate is 100 GJ tNH₃−1 which is cited as the efficiency level of a large scale Haber-Bosch process in 1930. https://www.sciencedirect.com/topics/engineering/haber-bosch-process
> 
> If my math is correct, that means your 8-charge (992g) ammonia recipe should cost about 99,200,000 J of energy (that's actually heat energy, meaning the electrical cost would be even higher), not 500.

> This is ignoring the problems of scaling (Haber-Bosch notoriously requires large reactors, high flow rates, high pressures, high temperatures, etc) that mean it's not at all practical to even drive this reaction electrically with the catalysts and gear you have outlined here and the fact that hydrogen electrolysis is also immensely inefficient without exotic catalysts.
> 
> At the end of the day, I'm expecting an overall conversion rate worse than 200 GJ / tNH3, meaning you're consuming well over 200 million battery charges per batch, as well as a much more robust and difficult crafting recipe for the machinery as well.

If going by that, it'd actually be 200 thousands, not millions (which is still more than the capacity of the largest storage battery, 80 thousands). According to description of "battery" item (the energy that is stored in batteries), each charge is assumed to be 1 kJ, not 1 J. I suppose I could further reduce the size of each batch, so that, in the end, 55000 of battery charges would make 2 units of ammonia, as I'm not entirely sure if multiple charged batteries would count towards this.

If that is the case, I'd also want to further modularize this machinery, making it possible to build a reactor which is heated up by the burning of coal rather than electricity, as acquiring enough of it may be easier than generating *this* much electricity. According to the article below, one kilogram of coal could provide 23.12 MJ of heat energy, therefore, assuming we could substitute 80% of energy needed with heat right away (correct me if it is a wrong assumption), it'd be down to 11000 battery charges used up (presumably for electrolysis and the operation of compressors) and 1.73 kg of coal burned up.
https://www.forbes.com/sites/startswithabang/2017/09/20/how-much-fuel-does-it-take-to-power-the-world/

This also allows to potentially integrate more advanced (and therefore energy-efficient) machinery into those recipes, such as using an electrolysis process which actually does make use of exotic catalysts like you mentioned for better energy efficiency.

As on the crafting recipe: my understanding is that for a less shoddy (and presumably actually working) variant of this machinery, built from things most likely salvaged from various industrial equipment, you'd need thicker-walled pipes (as well as pipe fittings), heavy-duty valves, pressure gauges, heavy-duty compressors (more powerful than what you'd find on a fridge), a built-in pyrometer and a pressure vessel as a reactor. Is there anything else I missed?

**Comment by kevingranade** [comment]
Yes I did get J and kJ mixed up, thanks, that brings it from "literally impossible" to "pointlessly expensive".

Note the *at least* in front of that 200MJ though, the 100MJ figure is based on something significantly larger and more robust than what the crafting recipes here imply, as well as methane-steam reformation for hydrogen production instead of electrolysis.

> I suppose I could further reduce the size of each batch, so that, in the end, 55000 of battery charges would make 2 units of ammonia, as I'm not entirely sure if multiple charged batteries would count towards this.

This would be ok with some additional sources outlining the efficiency of Hydrogen electrolysis.

From Wikipedia: A 100%-efficient electrolyser would consume 39.4 kilowatt-hours per kilogram (142 MJ/kg) of [hydrogen](https://en.wikipedia.org/wiki/Hydrogen),[[35]](https://en.wikipedia.org/wiki/Electrolysis_of_water#cite_note-EU140207-35) 12,749 joules per litre (12.75 MJ/m3). Practical electrolysis (using a [rotating electrolyser](https://en.wikipedia.org/wiki/Rotating_disk_electrode) at 15 bar pressure) may consume 50 kW⋅h/kg (180 MJ/kg), and a further 15 kW⋅h (54 MJ) if the hydrogen is compressed for use in hydrogen cars

A cobbled together hydrogen electrolysis system is going to have significantly worse efficiency than that. I haven't been able to find a source that directly states it, but I think you can derive it from the basic equations on wikipedia.

> If that is the case, I'd also want to further modularize this machinery, making it possible to build a reactor which is heated up by the burning of coal rather than electricity, as acquiring enough of it may be easier than generating this much electricity.

Let's focus on getting a relatively simple solution going, even if impractical, before branching out into other alternatives right off the bat.  At the end of the day, the *components* might be modular, but the final recipe needs to be "Haber process at X pressure and Y temperature using electrolysis for Hydrogen production and a molecular sieve for Nitrogen production".

> According to the article below, one kilogram of coal could provide 23.12 MJ of heat energy, therefore, assuming we could substitute 80% of energy needed with heat right away (correct me if it is a wrong assumption), it'd be down to 11000 battery charges used up (presumably for electrolysis and the operation of compressors) and 1.73 kg of coal burned up.

You're neglecting efficiencies, energy released is not the same as energy introduced captured by the process.  This also needs some kind of breakdown for what parts of the energy are doing what, not just an arbitrary split between heat and electricity inputs.

> Is there anything else I missed?

Proficiencies for all this stuff, it's not exactly common knowledge.

**Comment by AtomicFox556** [comment]
Updated as per request; info on power consumption of this equipment is now in comments in JSON. I've made an electrolyzer and the reactor separate items now, to potentially allow them to be replaced with better and/or differently-powered variants, or to be used for other processes should they ever be added.

**Comment by AtomicFox556** [comment]
Is there anything preventing this from being merged?

**Comment by I-am-Erk** [comment]
> Is there anything preventing this from being merged?

Sorry, this came at a time when we were all pretty busy. You currently have some conversations that need to be marked resolved if they are resolved, and some merge conflicts to fix. I'll be happy to take a closer look at it when those are done but at a brief glance I think it's good.

**Comment by AtomicFox556** [comment]
> Sorry, this came at a time when we were all pretty busy. You currently have some conversations that need to be marked resolved if they are resolved, and some merge conflicts to fix. I'll be happy to take a closer look at it when those are done but at a brief glance I think it's good.

It's all done.

**Comment by Zireael07** [comment]
dupe/related to [#57284](https://github.com/CleverRaven/Cataclysm-DDA/pull/57284) ?

**Pull request merged**